File: solutions/cpp/lasagna-master/1/lasagna_master.cpp
#include "lasagna_master.h"

namespace lasagna_master {

int preparationTime(std::vector<std::string> layers, int prep_time_per_layer){
    return layers.size()*prep_time_per_layer;
}
amount quantities(std::vector<std::string> layers){
    int layer_num=layers.size();
    int noodles=0, sauce=0;
    amount result;

    for(int layer_idx=0;layer_idx<layer_num;layer_idx++){
        std::string layer=layers[layer_idx];        
        if(layer=="noodles")noodles++;
        if(layer=="sauce")sauce++;
    }
    /*
    result={
        .noodles=noodles*50,
        .sauce=sauce*0.2,
    };
    */
    result.noodles=noodles*50;
    result.sauce=sauce*0.2;
    return result;
}
void addSecretIngredient(std::vector<std::string> &my_ingredients, const std::vector<std::string> &their_ingredients){
    int my_last_ingredient=my_ingredients.size()-1;
    int their_last_ingredient=their_ingredients.size()-1;

    my_ingredients[my_last_ingredient]=their_ingredients[their_last_ingredient];
}
std::vector<double> scaleRecipe(std::vector<double> amounts_two_portions, int portions){
    std::vector<double> result;
    int num_ingredients=amounts_two_portions.size();

    for(int ingredient=0;ingredient<num_ingredients;ingredient++){
        result.emplace_back(amounts_two_portions[ingredient]*portions/2);
    }
    return result;
}
void addSecretIngredient(std::vector<std::string> &my_ingredients, std::string her_ingredient){
    my_ingredients[my_ingredients.size()-1]=her_ingredient;
}
}  // namespace lasagna_master

File: solutions/javascript/lasagna-master/1/lasagna-master.js
/// <reference path="./global.d.ts" />
// @ts-check

/**
 * Implement the functions needed to solve the exercise here.
 * Do not forget to export them so they are available for the
 * tests. Here an example of the syntax as reminder:
 *
 * export function yourFunction(...) {
 *   ...
 * }
 */

export function cookingStatus(remainingMinutes){
  if(remainingMinutes===undefined){return 'You forgot to set the timer.';}
  if(remainingMinutes===0){return 'Lasagna is done.';}
  return 'Not done, please wait.';
}

export function preparationTime(layers, preparationMinutesPerLayer=2){
  return layers.length * preparationMinutesPerLayer;
}

export function quantities(layers){
  let noodles = 0;
  let sauce = 0;
  for(let idx=0;idx<layers.length;idx++){
    if (layers[idx]==='noodles'){noodles+=50;}
    if (layers[idx]==='sauce'){sauce+=0.2;}
  }
  return {noodles: noodles, sauce: sauce};
}

export function addSecretIngredient(friendsList, myList){
  myList.push(friendsList[friendsList.length-1]);
}

export function scaleRecipe(amountsForTwo, portions){
  let newAmounts = {};
  for(let ingredient in amountsForTwo){
    newAmounts[ingredient]=amountsForTwo[ingredient]*portions/2;
  }
  return newAmounts;
}
